utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/typed/AvroSchemaTypeDefinitionExtractor.scala

```diff
@@ -10,37 +10,37 @@ import java.nio.ByteBuffer
 import java.time.{Instant, LocalDate, LocalTime}
 import java.util.UUID
 
-/**
-  * Right now we're doing approximate type generation to avoid false positives in validation,
-  * so now we add option to skip nullable fields.
-  *
-  * @TODO In future should do it in another way
-  *
-  * @param skipOptionalFields
-  */
-class AvroSchemaTypeDefinitionExtractor(skipOptionalFields: Boolean) {
+object AvroSchemaTypeDefinitionExtractor {
 
   import collection.JavaConverters._
 
+  val DefaultPossibleTypes: Set[TypedClass] = Set(Typed.typedClass[GenericRecord])
+
+  val ExtendedPossibleTypes: Set[TypedClass] = DefaultPossibleTypes ++ Set(Typed.typedClass[java.util.Map[String, Any]])
+
+  val dictIdProperty = "nkDictId"
+
+  def typeDefinition(schema: Schema): TypingResult = typeDefinition(schema, DefaultPossibleTypes)
+
   /**
-    * see BestEffortAvroEncoder for underlying avro types
-    * !when applying changes keep in mind that this Schema.Type pattern matching is duplicated in {@link pl.touk.nussknacker.engine.avro.AvroDefaultExpressionDeterminer}
+    * See {@link pl.touk.nussknacker.engine.avro.encode.BestEffortAvroEncoder} for underlying avro types
+    *
+    * !When applying changes keep in mind that this Schema.Type pattern matching is duplicated in {@link pl.touk.nussknacker.engine.avro.AvroDefaultExpressionDeterminer},
+    * and is used at {@link  pl.touk.nussknacker.engine.avro.encode.AvroSchemaOutputValidator}
     */
   def typeDefinition(schema: Schema, possibleTypes: Set[TypedClass]): TypingResult = {
     schema.getType match {
       case Schema.Type.RECORD => {
         val fields = schema
           .getFields
           .asScala
-          //Field is marked as optional when field has default value
-          .filterNot(field => skipOptionalFields && field.hasDefaultValue)
           .map(field => field.name() -> typeDefinition(field.schema(), possibleTypes))
           .toList
 
         Typed(possibleTypes.map(pt => TypedObjectTypingResult(fields, pt)))
       }
-      case Schema.Type.ENUM =>  //It's should by Union, because output can store map with string for ENUM
-        Typed(Set(Typed.typedClass[EnumSymbol], AvroStringSettings.stringTypingResult))
+      case Schema.Type.ENUM =>
+        Typed.typedClass[EnumSymbol]
```

Comment:
It depends.. If you need string you have to do `.toString`. Before this change type enum hadn't worked properly: it was represented as Union[EnumSymbol, String], but when you did some comparing enum to string the result always was false, because in runtime there was always EnumType. Right now we are more strict, we allow to put string when avro contains enum but we don't allow comparing, use enum where string is required.